# Lab Assignment 2: How to Load CSV, ASCII, and other data into Python
## DS 6001: Practice and Application of Data Science

### Instructions
Please answer the following questions as completely as possible using text, code, and the results of code as needed. Format your answers in a Jupyter notebook. To receive full credit, make sure you address every part of the problem, and make sure your document is formatted in a clean and professional way.

There are 11 data files attached to this lab assignment, with different extensions. First, download all of these data files, and save them in the same folder on your local machine. Your task in the following questions is to load each file into Python correctly, so that you can begin the process of data cleaning. If the variable names are included in the file, use those names to name the columns. If the variable names are not included, use these names in order:

# km5ar
# keyu chen

In [1]:
column_names = ["Country", "Happiness score", "Whisker-high", "Whisker-low", 
  "Dystopia (1.92) + residual", "Explained by: GDP per capita", 
  "Explained by: Social support", "Explained by: Healthy life expectancy", 
  "Explained by: Freedom to make life choices", "Explained by: Generosity", 
  "Explained by: Perceptions of corruption" ]

If you loaded the data correctly, it will look like `data_clean.csv`, which is also attached to this lab.

## Problem 0
Import the libraries you will need. Then write code to change the working directory to the folder in which you saved the data files, run the code displayed above to create the `column_names` list, load `data_clean.csv`, and display the output of the `.info()` method of `data_clean`. (1 point)

In [2]:
# import the libraries
import pandas as pd
# change the working directory
%cd /Users/cory/Desktop/6001_Lab_2 #change to the dir you want 
#%pwd #- look #look at the current work dir 
p0 = pd.read_csv("data_clean.csv")
p0.info()
pd.set_option('display.max_columns',10)
pd.set_option('display.max_rows',10)


[Errno 2] No such file or directory: '/Users/cory/Desktop/6001_Lab_2 #change to the dir you want'
/Users/cory/Desktop/6001_Lab_2
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make l

## Problem 1
Load `data1.csv`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [3]:
p1 = pd.read_csv("data1.csv",skiprows=2)
p1.head() # check if we load the data correctly

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   

   Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                       2.595  ...                         1.592   
1                       2.383  ...                         1.582   
2                       2.370  ...                         1.590   
3                       2.426  ...                         1.644   
4                       2.320  ...                         1.549   

   Explained by: Healthy life expectancy  \
0                                  0.874   
1                                  0.861   
2                                  0.868   
3                                  0.914   
4                                  0.927   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357  

[5 rows x 11 columns]

We noticed the first two row of the data set is the Source information and URL of the data set, which is not useful for us, and we dicided to skip the first two row when we load the data by using skiprows = 2 in the pd.read_csv()

## Problem 2
Load `data2.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [4]:
p2= pd.read_csv("data2.txt", skiprows=2,comment = '/')
p2

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

due to the first two row are source and URL information, we decided to skip first two row, and we also find within the rows there are a lot of comment starts '/', so we use comment = '/' inside of our pd.read_csv to skip those comment

## Problem 3
Load `data3.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [5]:
p3 = pd.read_csv('data3.txt', skiprows = 2,sep = '\t')
p3

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

the first two rows is not useful for us, so we used skiprows = 2, then we also use sep = '\t' to seperate the data.

## Problem 4
Load `data4.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [6]:
p4 = pd.read_csv('data4.txt', header = None, sep = '$')
p4.columns = column_names
p4

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

since there is no header, we first set header = None, then we use sep = '$' to seperate each data 

## Problem 5
Load `data5.csv`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [7]:
p5 = pd.read_csv('data5.csv', skiprows = [157,158])
p5


Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

row 156, 157 have no value, so we use skiprows = [157,158] to remove it

## Problem 6
Load `data6.dat`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

In [8]:
p6 = pd.read_csv('data6.dat')
p6

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593      999.000   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                       999            3.355         3.448        3.262   
152                  Tanzania          999.000       999.000        3.193   
153               South Sudan            3.254       999.000        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074      999.000   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                       999.000   
1                       999.000  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                       999.000   
154                       2.487  ...                         0.000   
155                       1.752  ...                       999.000   

     Explained by: Healthy life expectancy  \
0                                  999.000   
1                                  999.000   
2                                  999.000   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                   999.000   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                   999.000   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                    999.000  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

## Problem 7
Load `data7.xlsx`, which is an Excel file. Keep only the sheet named “Data”. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

In [9]:
p7 = pd.read_excel(r'data7.xlsx', sheet_name='Data')
p7

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

since we only want to access the sheet "Data", we set sheet_name = 'Data' inside of pd.read_excel

## Problem 8
Load `data8.dta`, which is a Stata 13 file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

In [10]:
p8 = pd.read_stata('data8.dta')
p8.columns = column_names
p8

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

since this is a Stata 13 file, we decided to use pd.read_stat() to read in data, then we realized that the column names is not capitalized, so we renamed it by using p8.columns


## Problem 9
Load `data9.sav`, which is an SPSS file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

In [11]:
p9 = pd.read_spss('data9.sav')
p9.columns = column_names
p9

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

since this is a SPSS file, then i used pd.read_spss() to read in the data, then the first letter is capitalized, so i renamed it by using p9.columns = column_names 

## Problem 10
Load `data10.xpt`, which is a SAS file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (If some of the country names display as `b'Finland'`, don't worry aout that.) (2 points)

In [12]:
p10 = pd.read_sas('data10.xpt')
p10.columns = column_names
p10

Country  Happiness score  Whisker-high  Whisker-low  \
0                     b'Finland'            7.632         7.695        7.569   
1                      b'Norway'            7.594         7.657        7.530   
2                     b'Denmark'            7.555         7.623        7.487   
3                     b'Iceland'            7.495         7.593        7.398   
4                 b'Switzerland'            7.487         7.570        7.405   
..                           ...              ...           ...          ...   
151                     b'Yemen'            3.355         3.448        3.262   
152                  b'Tanzania'            3.303         3.414        3.193   
153               b'South Sudan'            3.254         3.385        3.123   
154  b'Central African Republic'            3.083         3.227        2.939   
155                   b'Burundi'            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                  1.592000e+00   
1                         2.383  ...                  1.582000e+00   
2                         2.370  ...                  1.590000e+00   
3                         2.426  ...                  1.644000e+00   
4                         2.320  ...                  1.549000e+00   
..                          ...  ...                           ...   
151                       1.106  ...                  1.073000e+00   
152                       0.628  ...                  9.910000e-01   
153                       1.691  ...                  6.080000e-01   
154                       2.487  ...                  5.397605e-79   
155                       1.752  ...                  6.270000e-01   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

due to this is a SAS file, I chose to use pd.read_sas() to read in the data, and I realized that the column name started "explianed" isn't correct, so i used p10.columns = column_names to correct the error

## Problem 11
Please load the `data11.txt` file, which is a fixed width file. The columns are defined as follows:

| Variable                                   | Width | Start | End |
|--------------------------------------------|-------|-------|-----|
| Country                                    | 24    | 1     | 24  |
| Happiness score                            | 5     | 25    | 29  |
| Whisker-high                               | 5     | 30    | 34  |
| Whisker-low                                | 5     | 35    | 39  |
| Dystopia (1.92) + residual                 | 5     | 40    | 44  |
| Explained by: GDP per capita               | 5     | 45    | 49  |
| Explained by: Social support               | 5     | 50    | 54  |
| Explained by: Healthy life expectancy      | 5     | 55    | 59  |
| Explained by: Freedom to make life choices | 5     | 60    | 64  |
| Explained by: Generosity                   | 5     | 65    | 69  |
| Explained by: Perceptions of corruption    | 5     | 70    | 74  |

Then save the this loaded data frame as a CSV file on your local machine. Be sure to use a unique filename so as not to overwrite any existing files. (5 points)

In [13]:
p11 = pd.read_fwf("data11.txt",header = None,sep = '.')

In [14]:
datanames = ['Country','Happiness score','Whisker-high','Whisker-low','Dystopia (1.92) + residual','Explained by: GDP per capita','Explained by: Social support','Explained by: Healthy life expectancy','Explained by: Freedom to make life choices','Explained by: Generosity','Explained by: Perceptions of corruption']
datawidths = [24,5,5,5,5,5,5,5,5,5,5]
datapos = [[0,24], [25,29], [30,34], [35,39], [40,44],
    [45,49], [50,54], [55,59], [60,64], [65,69],
    [70,74]]

In [15]:
data11 =pd.read_fwf("data11.txt", widths=datawidths, header=None, names=datanames)
data11

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            7.632         7.695        7.569   
1                      Norway            7.594         7.657        7.530   
2                     Denmark            7.555         7.623        7.487   
3                     Iceland            7.495         7.593        7.398   
4                 Switzerland            7.487         7.570        7.405   
..                        ...              ...           ...          ...   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         2.595  ...                         1.592   
1                         2.383  ...                         1.582   
2                         2.370  ...                         1.590   
3                         2.426  ...                         1.644   
4                         2.320  ...                         1.549   
..                          ...  ...                           ...   
151                       1.106  ...                         1.073   
152                       0.628  ...                         0.991   
153                       1.691  ...                         0.608   
154                       2.487  ...                         0.000   
155                       1.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

In [16]:
data11_2 = pd.read_fwf("data11.txt", colspecs=datapos, header=None, names=datanames)
data11_2

Country  Happiness score  Whisker-high  Whisker-low  \
0                     Finland            0.632         0.695        0.569   
1                      Norway            0.594         0.657        0.530   
2                     Denmark            0.555         0.623        0.487   
3                     Iceland            0.495         0.593        0.398   
4                 Switzerland            0.487         0.570        0.405   
..                        ...              ...           ...          ...   
151                     Yemen            0.355         0.448        0.262   
152                  Tanzania            0.303         0.414        0.193   
153               South Sudan            0.254         0.385        0.123   
154  Central African Republic            0.083         0.227        0.939   
155                   Burundi            0.905         0.074        0.735   

     Dystopia (1.92) + residual  ...  Explained by: Social support  \
0                         0.595  ...                         0.592   
1                         0.383  ...                         0.582   
2                         0.370  ...                         0.590   
3                         0.426  ...                         0.644   
4                         0.320  ...                         0.549   
..                          ...  ...                           ...   
151                       0.106  ...                         0.073   
152                       0.628  ...                         0.991   
153                       0.691  ...                         0.608   
154                       0.487  ...                         0.000   
155                       0.752  ...                         0.627   

     Explained by: Healthy life expectancy  \
0                                    0.874   
1                                    0.861   
2                                    0.868   
3                                    0.914   
4                                    0.927   
..                                     ...   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                         0.681                     0.192   
1                                         0.686                     0.286   
2                                         0.683                     0.284   
3                                         0.677                     0.353   
4                                         0.660                     0.256   
..                                          ...                       ...   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
0                                      0.393  
1                                      0.340  
2                                      0.408  
3                                      0.138  
4                                      0.357  
..                                       ...  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  

[156 rows x 11 columns]

In [17]:
data11_2.to_csv('output_labassignment2.csv')

due to the data11.txt is a fixed width file, and at the same time we also have the relative information of the fixed width file, such as variable names, width, start and end data. we chose to use two way to read in data, the first is read in the data by using datanames and data width, and the second way is use data names and start and end data. We realized when we use the start and end data, the start for country should 0 instead of 1 provided from the table. so we changed 1 to 0. 
Finally we output it into csv file.